src/components/Context.js

```diff
@@ -70,10 +74,11 @@ class ProductProvider extends Component {
     discountedAmount: 0,
   };
 
-  dispatch = (type) => {
+  dispatch = (type, data) => {
```

Comment:
**here's the key trick**

you don't need to:
```
setState({
   some shit
},
() => {
  dispatch(...)
}
```

dispatch is meant to update the state, and you can pass payload along with it